conqmetad.man
.\" $Id$
.TH "conqmetad" 6 "" ""
.SH NAME
conqmetad \- a Conquest Meta server
.SH SYNOPSIS
.PP
\fBconqmetad\fP [OPTIONS]
.SH DESCRIPTION
.PP
\fBconqmetad\fP is a meta server for Conquest.  A meta server listens
for game availablity announcements from other \fBconquestd\fP servers
scattered about, and makes this list available to conquest clients that use the
\fB\-m\fP option.  This makes it easy for people to find prospective
games. 
.PP
The de-facto Conquest Meta server runs at \fBconquest.radscan.com\fP.
.PP
The meta server listens on port UDP port 1700 for announcements from
\fBconquestd\fP servers.  TCP connections to port 1700 cause
\fBconqmetad\fP to dump the current server list and their associated
data in an easily parsable text format, one server per line, and
disconnect. 
.SH "OPTIONS"
.PP
.TP 8
.B \-d 
Daemon mode.  
.TP 8
.B \-l
Listen for local connections only.  When this option is supplied, only
clients on the local host will be able to connect.
.TP 8
.B \-u <user>
Run as user \fIuser\fP.  
.TP 8
.B \-v
Be more verbose.  
.SH "AUTHOR"
Jon Trulson
.SH "SEE ALSO"
.PP
conquest(6), conquestd(6), conqoper(6), conqdriv(6), conqmetad(6) 
.PP
See README in the conquest distribution (usually installed in
<conquest install prefix>/etc/) for information and tips on how to
play. 
.PP
See INSTALL for instructions on unpacking, compiling, and installing
conquest, as well as information on terminfo requirements (for the
curses client) and other system specific issues.
.PP
See README.cygwin for information on building/using conquest on MS
Windows systems. 
.PP
See HISTORY for the latest news and modification history.
.PP
See CONQINIT.TXT for information on CQI, and the format of the
conqinitrc and texturesrc files.
.SH "COPYRIGHT"
.PP
Copyright (C)1994-2017 by Jon Trulson <jon@radscan.com> under the MIT
License.
.PP
Original VAX/VMS Ratfor Conquest:
.PP
Copyright (C)1983-1986 by Jef Poskanzer and Craig Leres.  Permission to
use, copy, modify, and distribute this software and its documentation
for any purpose and without fee is hereby granted, provided that this
copyright notice appear in all copies and in all supporting
documentation. Jef Poskanzer and Craig Leres make no representations
about the suitability of this software for any purpose. It is provided
"as is" without express or implied warranty.


HISTORY

         +--------------------------------------------------------+
         |  CCC    OOO   N   N   QQQ   U   U  EEEEE   SSSS  TTTTT |
         | C   C  O   O  NN  N  Q   Q  U   U  E      S        T   |
         | C      O   O  N N N  Q   Q  U   U  EEE     SSS     T   |
         | C   C  O   O  N  NN  Q  Q   U   U  E          S    T   |
         |  CCC    OOO   N   N   QQ Q   UUU   EEEEE  SSSS     T   |
         +--------------------------------------------------------+

                              Conquest HISTORY


8.5.1 (stable, windows only) 4/22/2010

  - add a call the WSAStartup() to conquestgl so socket i/o can
    *consistently* work :)

8.5 (stable) 04/20/2010

  - A new stable release, at long last.

  - Major highlights of this release, compared to the last stable
    release (8.4.1 09/13/2008):

    - General:

      - MAC OSX (Darwin) support.  Full sound and hardware accelerated
        OpenGL is supported.  Note, you will probably need to get SDL
        and SDL_mixer, as well as libogg and lib vorbis in order to
        build.  

      - Native Windows XP/Windows 7 support for the OpenGL client
        (conquestgl) using the MinGW cross-compiler enviroment.  Full
        sound and hardware accelerated OpenGL is supported.  A zip
        file will be available on sourceforge containing a
        pre-compiled conquestgl binary, and support files.  See
        README.Windows (or Conquest\INSTALL.TXT in the ZIP file) for
        more information on installtion and use.

      - a re-work of packet handling.  The original goal was to
        eventually support non-blocking socket IO, however some
        significant (and incompatible) protocol changes will be
        required to do this properly.  Next major release will tackle
        this. 

      - rework of texture and color representation internally.

      - remove telnet (-t) support from the clients.  Nobody runs
        telnet conquest games anymore.

      - switched to Mercurial (hg) for source code control.

        I will 'hg tag' each devel release.  For stable releases (like
        this one), a branch name will be created instead of a tag.

        Current development will always happen at the tip, in the
        'default' branch, and may therefore be unstable.

        Bug fixes and enhancements made to 8.5 will occur in the
        conquest-8.5 branch.

        A command like (all on one line):

          hg clone -u conquest-8.5 \
             http://conq.hg.sourceforge.net/hgroot/conq/conq conquest

          can be used to check out the stable branch for version 8.5.
          Omit the '-u conquest-8.5' if you want the current state of
          development.

    - Server fixes:

      - fix server bug that would cause vacant ships, when destroyed,
        to fail to explode properly, leaving some stationary torps
        behind.
      
      - correct a server problem with UDP keepalive packets.  During
        certain operations (like beaming, bombing, self destruct, etc)
        these packets could improperly terminate these activities on
        the server, leaving the client to believe these activities
        were still in progress, causing great angst in the land.

    - Client fixes (primarily the OpenGL client, conquestgl):

      - New Orion ship textures (thanks to blender and the fine
        contributors at ShareGC.com).  See the CREDITS file.

      - some changes to work better with wide-screen display formats

      - new ship shield textures, both for the icon-ship display and
        the main viewer.

      - in conquestgl, torps are no longer colored according to their
        'danger' level.  It's up to you, as the ship's Captain, to
        know when a torp may kill you.  When in doubt, raise
        shields and run like the coward you are. :)

      - rework of the icon-ship display:

        - now use a luminance variant of the main ship texture as the
          ico-ship rather than a seperate texture.

        - implement 'lamps' rather than the text (cloaking, repair,
          shields, etc)

        - make the ico-ship drawing area square to avoid stretching
          issues

        - use ship's torp texture for the torp-pips

        - use ship's phaser texture for the phaser recharge indicator

      - fix bug that could cause the use of undefined mouse macros to
        improperly terminate certain operations (beaming, bombing,
        etc). 

  - changes from the last devel (8.4.1i) release:

    - Fix up some documentation for Windows systems.  Conquestgl has
      been tested on Windows 7, seems to work great.

    - fix up stupidness in handling cInfo.remotehost data (caused
      invalid free()'s, depending on options specified).

    - Do not declare a crypt prototype on DARWIN systems

    - Move 'q' from beginning of line in conquest.man, lintian error
      reported by Jonathon (new debian maintainer)

    - Remove global.c, instantiate global.h and conf.h within all of
      the <main>.c files.  Needed by Darwin systems to avoid linker
      errors when building conqstrat and conqinit executables.

    - Add some icon pngs and a .ico for windows


  ## The following are the devel release notes for more detail.  Use
     'hg log -v' on the repository for even more detail if you care.

  8.4.1i (devel) 03/27/2010

  - unless there is a big problem, a 'stable' official release will be
    up next.  

  - Switch to using Mercurial (hg) for source code control.  I totally
    love it compared to svn.

  - MAC OS X (Darwin) support.  Note this support is still a little
    fuzzy, in regards to counquest group creation, etc.  I do not have
    direct access to a Darwin system, so this port was accomplished by
    back-and-forth with a friend who does.  Once you can get it built,
    OpenGL and sound is fully functional.

  - Windows (XP) native support via MinGW cross-compiler toolchain :)
    See README.Windows for installation instructions (or INSTALL.TXT

conqmetad.c
#include "c_defs.h"

/************************************************************************
 *
 * conqmetad - the Conquest meta server.  Similiar to conquestd no? :)
 *
 * Copyright Jon Trulson under the MIT License. (See LICENSE).
 ***********************************************************************/

#define NOEXTERN_GLOBALS
#include "global.h"

#include "cb.h"

#define NOEXTERN_CONTEXT
#include "context.h"

#include "conqdef.h"
#include "conqnet.h"
#include "conqunix.h"

#define NOEXTERN_CONF
#include "conf.h"

#include "conqutil.h"

#include "meta.h"

#include "tcpwrap.h"

#define LISTEN_BACKLOG 5 /* # of requests we're willing to to queue */

metaSRec_t metaServerList[META_MAXSERVERS];
static uint16_t listenPort = META_DFLT_PORT;
static char *progName;
static int localOnly = FALSE;   /* whether to only listen on loopback */
static int maxSlot = 0;         /* the max slot we've ever used */
const int expireSeconds = (60 * 5); /* 5 minutes */

void catchSignals(void);
void handleSignal(int sig);

static int cmpmeta(void *cmp1, void *cmp2)
{
    register int *icmp1, *icmp2;

    icmp1 = (int *) cmp1;
    icmp2 = (int *) cmp2;

    /* we sort based on active ships */

    if (metaServerList[*icmp1].numactive > metaServerList[*icmp2].numactive)
        return(-1);
    else if (metaServerList[*icmp1].numactive < metaServerList[*icmp2].numactive)
        return(1);
    else
        return(0);
}


static void sortmetas( int mv[], int numentries )
{
    qsort(mv, numentries, sizeof(int),
          (int (*)(const void *, const void *))cmpmeta);

    return;

}

void printUsage()
{
    printf("Usage: conqmetad [ -d ] [ -l ] [ -u user ] [ -v ] \n");
    printf("   -d            daemon mode\n");
    printf("   -l            listen for local connections only\n");
    printf("   -u user       run as user 'user'.\n");
    printf("   -v            be more verbose.\n");

    return;
}

void ageServers(void)
{
    int i;
    time_t now = time(0);

    for (i=0; i<maxSlot; i++)
    {
        if (metaServerList[i].valid)
            if (abs(metaServerList[i].lasttime - now) > expireSeconds)
            {
                utLog("META: expiring %s:%u(%s)\n",
                      metaServerList[i].altaddr,
                      metaServerList[i].port,
                      metaServerList[i].addr);
                metaServerList[i].valid = FALSE;
            }
    }

    return;
}


/* find the server slot for this record (or make a new one).  returns the
   slot number found/'created', or -1 for error */
int findSlot(metaSRec_t *srec, int *isupdate)
{
    int i, found;
    int rv = 0;

    *isupdate = FALSE;

    /* first look for it */
    found = FALSE;
    for (i=0; i<maxSlot; i++)
    {
        if ((!strcmp(metaServerList[i].addr, srec->addr)) &&
            (!strcmp(metaServerList[i].altaddr, srec->altaddr)) &&
            metaServerList[i].valid && (metaServerList[i].port == srec->port))
        {
            rv = i;
            found = TRUE;
            *isupdate = TRUE;
            break;
        }
    }

    if (!found)
    {                           /* didn't find one */
        for (i=0; i<META_MAXSERVERS; i++)
            if (!metaServerList[i].valid)
            {                     /* found a slot */
                rv = i;
                if (rv >= maxSlot)
                    maxSlot = rv + 1;
                found = TRUE;
                break;
            }
    }

    if (!found)
        return -1;                  /* nothing there george */
    else
        return rv;
}


void initServerList()
{
    int i;

    memset((void *)&metaServerList, 0, sizeof(metaSRec_t) * META_MAXSERVERS);

    for (i=0; i<META_MAXSERVERS; i++)
        metaServerList[i].valid = FALSE;

    return;
}

void metaProcList(int sock, char *hostbuf)
{
    static int mvec[META_MAXSERVERS];
    int nm;
    char tbuf[BUFFERSZ];
    int i;

    /* init mvec */
    nm = 0;
    for (i=0; i < maxSlot; i++)
        if (metaServerList[i].valid)
            mvec[nm++] = i;

    sortmetas(mvec, nm);

    /* dump the sorted server list */
    for (i=0; i<nm; i++)
    {
        metaServerRec2Buffer(tbuf, &metaServerList[mvec[i]]);
        if (write(sock, tbuf, strlen(tbuf)) <= 0)
            utLog("META: write failed to %s", hostbuf);
    }

    utLog("META: server query from %s", hostbuf);

    return;
}

void metaProcUpd(char *buf, int rlen, char *hostbuf)
{
    metaSRec_t sRec;
    int slot;
    int wasfound;

    if (!metaBuffer2ServerRec(&sRec, buf))
    {
        utLog("META: malformed buffer '%s', ignoring", buf);
        return;
    }

    utStrncpy(sRec.addr, hostbuf, CONF_SERVER_NAME_SZ);

    /* if altaddr is empty, we copy hostbuf into it. */

    if (strlen(sRec.altaddr) == 0)
        utStrncpy(sRec.altaddr, sRec.addr, CONF_SERVER_NAME_SZ);

    /* now find a slot for it. */
    if ((slot = findSlot(&sRec, &wasfound)) == -1)
    {
        utLog("META: findSlot couldn't find one, ignoring\n");
        return;
    }

    /* init the slot */
    sRec.valid = TRUE;
    sRec.lasttime = time(0);
    if (sRec.port == 0)
        sRec.port = CN_DFLT_PORT;
    metaServerList[slot] = sRec;

    if (!wasfound)                /* new server */
        utLog("META: Added server %s:%u(%s), slot %d",
              metaServerList[slot].altaddr,
              metaServerList[slot].port,
              metaServerList[slot].addr,
              slot);
    else
    {
#if defined(DEBUG_META)
        utLog("META: Updated server %s(%s)",
              metaServerList[slot].altaddr,
              metaServerList[slot].addr);
#endif
    }

    return;
}

/* we exit here on serious errors */
/* we listen on both a UDP and TCP socket.  UDP is used to register a server
   TCP is used to get a dump of the current servers. */
void metaListen(void)
{
    int s, t, tc;			/* socket descriptor */
    struct sockaddr_in sa, isa;	/* internet socket addr. structure UDP */
    struct sockaddr_in tsa, tisa;	/* internet socket addr. structure TCP */
    struct timeval tv;
    struct hostent *hp;
    int rv, rlen;
    socklen_t sockln;
    char hostbuf[CONF_SERVER_NAME_SZ];
    char rbuf[BUFFERSZ];
    socklen_t alen;
    fd_set readfds;

    sa.sin_port = htons(listenPort);
    tsa.sin_port = htons(listenPort);

    if (localOnly)
    {
        sa.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
        tsa.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    }
    else
    {
        sa.sin_addr.s_addr = htonl(INADDR_ANY);
        tsa.sin_addr.s_addr = htonl(INADDR_ANY);
    }

    sa.sin_family = AF_INET;
    tsa.sin_family = AF_INET;

    /* allocate an open socket for incoming UDP connections */
    if (( s = socket(AF_INET, SOCK_DGRAM, 0)) < 0)
    {
        perror ( "UDP socket" );
        exit(1);
    }

    /* allocate an open socket for incoming TCP connections */
    if (( t = socket(AF_INET, SOCK_STREAM, 0)) < 0)
    {
        perror ( "TCP socket" );
        exit(1);
    }

    /* bind the socket to the service port so we hear incoming
     * connections
     */
    if ( bind( s, (struct sockaddr *)&sa, sizeof ( sa )) < 0 )
    {
        perror( "UDP bind" );
        exit(1);
    }

    if ( bind( t, (struct sockaddr *)&tsa, sizeof ( tsa )) < 0 )
    {
        perror( "TCP bind" );
        exit(1);
    }

    /* set the maximum connections we will fall behind */
    listen( t, LISTEN_BACKLOG );

    alen = sizeof(struct sockaddr);

    utLog("NET: meta server listening on TCP and UDP port %d\n", listenPort);

    /* go into infinite loop waiting for new connections */
    while (TRUE)
    {
        tv.tv_sec = 30;           /* age servers every 30 secs */
        tv.tv_usec = 0;
        FD_ZERO(&readfds);
        FD_SET(s, &readfds);
        FD_SET(t, &readfds);

        if ((rv = select(max(s,t)+1, &readfds, NULL, NULL, &tv)) < 0)
        {
            utLog("META: select failed: %s", strerror(errno));
            exit(1);
        }

        /* TCP (list) socket */
        if (FD_ISSET(t, &readfds))
        {
            sockln = sizeof (isa);
            if ((tc = accept(t, (struct sockaddr *)&tisa, &sockln )) < 0)
            {
                utLog("META: accept failed: %s", strerror(errno) );
                continue;
            }

            if ((hp = gethostbyaddr((char *) &tisa.sin_addr.s_addr,
                                    sizeof(unsigned long),
                                    AF_INET)) == NULL)
            {
                utStrncpy(hostbuf, inet_ntoa((struct in_addr)tisa.sin_addr),
                          CONF_SERVER_NAME_SZ);
            }
            else
            {
                utStrncpy(hostbuf, hp->h_name, CONF_SERVER_NAME_SZ);
            }

            hostbuf[CONF_SERVER_NAME_SZ - 1] = 0;

            if (!tcpwCheckHostAccess(TCPW_DAEMON_CONQMETAD, hostbuf))
            {
                close(tc);
                continue;
            }

            metaProcList(tc, hostbuf);
            close(tc);
        }

        /* UDP (ping) socket */
        if (FD_ISSET(s, &readfds))
        {
            memset(rbuf, 0, BUFFERSZ);
            rlen = recvfrom(s, rbuf, BUFFERSZ, 0,
                            (struct sockaddr *)&isa, &alen);

            if ((hp = gethostbyaddr((char *) &isa.sin_addr.s_addr,
                                    sizeof(unsigned long),
                                    AF_INET)) == NULL)
            {
                utStrncpy(hostbuf, inet_ntoa((struct in_addr)isa.sin_addr),
                          CONF_SERVER_NAME_SZ);
            }
            else
            {
                utStrncpy(hostbuf, hp->h_name, CONF_SERVER_NAME_SZ);
            }

            hostbuf[CONF_SERVER_NAME_SZ - 1] = 0;

            if (!tcpwCheckHostAccess(TCPW_DAEMON_CONQMETAD, hostbuf))
            {
                continue;
            }

            metaProcUpd(rbuf, rlen, hostbuf);

        }

        /* timeout */
        ageServers();
    }

    return;			/* NOTREACHED */
}


/*  conqmetad - main program */
int main(int argc, char *argv[])
{
    int i;
    char *myuidname = NULL;              /* what user do I run under? */
    int dodaemon = FALSE;

    progName = argv[0];

    while ((i = getopt(argc, argv, "dlp:u:v")) != EOF)    /* get command args */
        switch (i)
        {
        case 'd':
            dodaemon = TRUE;
            break;

        case 'p':
            listenPort = (uint16_t)atoi(optarg);
            break;

        case 'l':                 /* local conn only */
            localOnly = TRUE;
            break;

        case 'u':
            myuidname = optarg;
            break;

        case 'v':
            cqDebug++;
            break;

        default:
            printUsage();
            exit(1);
        }

    if ((ConquestGID = getConquestGID()) == -1)
    {
        fprintf(stderr, "%s: getConquestGID() failed\n", progName);
        exit(1);
    }


    /* at this point, we see if the -u option was used.  If it was, we
       setuid() to it */

    if (myuidname)
    {
        int myuid;

        if ((myuid = getUID(myuidname)) == -1)
        {
            fprintf(stderr, "%s: getUID(%s) failed\n", progName, myuidname);
            exit(1);
        }

        if (setuid(myuid) == -1)
        {
            fprintf(stderr, "%s: setuid(%d) failed: %s\n", progName, myuid,
                    strerror(errno));
            exit(1);
        }
        else
            utLog("INFO: META running as user '%s', uid %d.", myuidname, myuid);
    }


#ifdef DEBUG_CONFIG
    utLog("%s@%d: main() Reading Configuration files.", __FILE__, __LINE__);
#endif

    if (GetSysConf(FALSE) == -1)
    {
#ifdef DEBUG_CONFIG
        utLog("%s@%d: main(): GetSysConf() returned -1.", __FILE__, __LINE__);
#endif
    }

    /* if daemon mode requested, fork off and detach */
    if (dodaemon)
    {
        int cpid;
        utLog("INFO: becoming daemon");
        if (chdir("/") == -1)
        {
            utLog("chdir(/) failed: %s", strerror(errno));
            exit(1);
        }



        cpid = fork();
        switch (cpid)
        {
        case 0:
            /* child */

#if defined(HAVE_DAEMON)
            if (daemon(0, 0) == -1)
            {
                utLog("daemon() failed: %s", strerror(errno));
                exit(1);
            }

#else
# if defined(HAVE_SETPGRP)
#  if defined(SETPGRP_VOID)
            setpgrp ();
#  else
            setpgrp (0, getpid());
#  endif
# endif

            close (0);
            close (1);
            close (2);

            /* Set up the standard file descriptors. */

            (void) open ("/", O_RDONLY);        /* root inode already in core */
            (void) dup2 (0, 1);
            (void) dup2 (0, 2);

#endif /* !HAVE_DAEMON */

            break;

        case -1:
            /* error */
            fprintf(stderr, "daemon fork failed: %s\n", strerror(errno));
            break;

        default:
            /* parent */
            exit(0);
        }
    }

    initServerList();

    /* setup, listen for, and process  client connections. */

    metaListen();

    exit(0);

}

void catchSignals(void)
{
    signal(SIGHUP, (void (*)(int))handleSignal);
    signal(SIGTSTP, SIG_IGN);
    signal(SIGTERM, (void (*)(int))handleSignal);
    signal(SIGINT, SIG_IGN);
    signal(SIGQUIT, (void (*)(int))handleSignal);
    signal(SIGCLD, SIG_IGN);	/* allow children to die */

    return;
}


void handleSignal(int sig)
{
    utLog("META: exiting on signal %d", sig);
    exit(0);
}
